src/replication/adapter/helmhub/client.go

```diff
@@ -0,0 +1,116 @@
+package helmhub
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/goharbor/harbor/src/replication/model"
+	"github.com/goharbor/harbor/src/replication/util"
+	"github.com/pkg/errors"
+	"io/ioutil"
+	"net/http"
+)
+
+// ErrHTTPNotFound defines the return error when receiving 404 response code
+var ErrHTTPNotFound = errors.New("Not Found")
+
+// Client is a client to interact with HelmHub
+type Client struct {
+	client *http.Client
+}
+
+// NewClient creates a new HelmHub client.
+func NewClient(registry *model.Registry) *Client {
+	return &Client{
+		client: &http.Client{
+			Transport: util.GetHTTPTransport(false),
+		},
+	}
+}
+
+// fetchCharts fetches the chart list from helm hub.
+func (c *Client) fetchCharts() (*chartList, error) {
+	request, err := http.NewRequest(http.MethodGet, baseURL+listCharts, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("fetch chart list error %d: %s", resp.StatusCode, string(body))
+	}
+
+	list := &chartList{}
+	err = json.Unmarshal(body, list)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal chart list response error: %v", err)
+	}
+
+	return list, nil
+}
+
+// fetchChartDetail fetches the chart detail of a chart from helm hub.
+func (c *Client) fetchChartDetail(chartName string) (*chartVersionList, error) {
+	request, err := http.NewRequest(http.MethodGet, baseURL+listVersions(chartName), nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := c.client.Do(request)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNotFound {
+		return nil, fmt.Errorf("fetch chart detail error %d: %s", resp.StatusCode, string(body))
+	} else if resp.StatusCode == http.StatusNotFound {
+		return nil, ErrHTTPNotFound
+	}
+
+	list := &chartVersionList{}
+	err = json.Unmarshal(body, list)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshal chart detail response error: %v", err)
+	}
+
+	return list, nil
+}
+
+func (c *Client) checkHealthy() error {
+	request, err := http.NewRequest(http.MethodGet, baseURL, nil)
```

Comment:
Shall we check the` index.yaml`  service endpoint? Seems the base URL is the root URL of the website?

Reply:
i don't know what you mean. but i couldn't find any related API. and i want to use `/charts` as health URL at first, but response body is too large.